packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,558 @@
+#include "packager/VisibilityChecker.h"
+#include "ast/treemap/treemap.h"
+#include "common/concurrency/ConcurrentQueue.h"
+#include "common/formatting.h"
+#include "common/sort.h"
+#include "core/Context.h"
+#include "core/errors/packager.h"
+#include "core/errors/resolver.h"
+#include <iterator>
+
+namespace sorbet::packager {
+
+namespace {
+
+// For each __package.rb file, traverse the resolved tree and apply the visibility annotations to the symbols.
+class PropagateVisibility final {
+    const core::packages::PackageInfo &package;
+
+    bool definedByThisPackage(const core::GlobalState &gs, core::ClassOrModuleRef sym) {
+        auto pkg = gs.packageDB().getPackageNameForFile(sym.data(gs)->loc().file());
+        return this->package.mangledName() == pkg;
+    }
+
+    void recursiveExportSymbol(core::GlobalState &gs, bool firstSymbol, core::ClassOrModuleRef klass) {
+        // There is a loop between a class and its singleton, and this is the easiest way to detect it.
+        if (!firstSymbol && klass.data(gs)->flags.isExported) {
+            return;
+        }
```

Comment:
do we have a test for this, ooc?

Reply:
The loop manifests as non-termination without that, but we don't have an explicit test for the part that relies on `firstSymbol`: exporting a child of a class before exporting the parent. I can add a test for that.